walletlibrary/src/main/java/com/microsoft/walletlibrary/did/sdk/datasource/network/BaseNetworkOperation.kt

```diff
@@ -29,52 +30,56 @@ import java.io.IOException
  * In default methods, S == T, for no transformation takes place.
  * fire method will just return Result.Success(responseBody: S)
  */
-internal abstract class BaseNetworkOperation<S, T> {
+internal abstract class BaseNetworkOperation<T> {
 
-    abstract val call: suspend () -> Response<S>
+    abstract val call: suspend () -> Result<IResponse>
 
-    open suspend fun fire(): Result<T> {
+    abstract suspend fun toResult(response: IResponse): Result<T>
+
+    suspend inline fun fire(): Result<T> {
         try {
-            val response = logNetworkTime("${this::class.simpleName}") {
+            logNetworkTime("${this::class.simpleName}") {
                 call.invoke()
+            }.onSuccess {
+                return toResult(it)
+            }.onFailure {
+                return onFailure(it)
             }
-            if (response.isSuccessful) {
-                return onSuccess(response)
-            }
-            return onFailure(response)
         } catch (exception: IOException) {
-            return Result.Failure(LocalNetworkException("Failed to send request.", exception))
+            return Result.failure(LocalNetworkException("Failed to send request.", exception))
         }
-    }
-
-    open suspend fun onSuccess(response: Response<S>): Result<T> {
-        // TODO("how do we want to handle null bodies")
-        // TODO("how to not suppress this warning")
-        @Suppress("UNCHECKED_CAST")
-        val transformedPayload = (response.body() ?: throw LocalNetworkException("Body of Response is null.")) as T
-        return Result.Success(transformedPayload)
+        return Result.failure(SdkException("Failed to get a response"))
     }
 
     // TODO("what do we want our base to look like")
-    open fun onFailure(response: Response<S>): Result<Nothing> {
-        val responseBody = response.errorBody()?.string() ?: ""
+    open fun onFailure(exception: Throwable): Result<T> {
```

Comment:
For failure cases, it wouldn't return <T> rather a failure result. Was this changed for a specific reason?

Reply:
Nope. It also just calls .failure so it could remain Result<Nothing>.